File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def dfs(R,C,row,col,board,word,index):
            if(len(word)==index):
                return True
            
            directions=[[-1,0],[1,0],[0,1],[0,-1]]
            for dr,dc in directions:
                adjr=row+dr
                adjc=col+dc
                if(adjr<R and adjr>=0 and adjc<C and adjc>=0 and visit[adjr][adjc]==0 and board[adjr][adjc]==word[index]):
                    visit[adjr][adjc]=1
                    flag=dfs(R,C,adjr,adjc,board,word,index+1)
                    if(flag):
                        return True
            visit[row][col]=0
            return False

                    






        visit=[]
        R,C=len(board),len(board[0])
        #initialising the flag array
        for row in range(R):
            visit.append([])
            for col in range(C):
                visit[row].append(0)
        #Traversing the array
        index=0
        for row in range(R):
            for col in range(C):
                if(board[row][col]==word[index]):#checking the start of the word in the board
                    visit[row][col]=1
                    found=dfs(R,C,row,col,board,word,index+1)
                    if(found):
                        return True
        return False
        

File: 0079-word-search/0079-word-search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        int R=board.length;
        int C=board[0].length;
        boolean visited[][]=new boolean[R][C];
        for(int r=0;r<R;r++){
            for(int c=0;c<C;c++){
                if(word.charAt(0)==board[r][c]){
                    if(search(r,c,R,C,board,word,0,visited)){
                        return true;
                    }
                }
            }
        }
        return false;
    }
    private static boolean search(int s,int e,int R,int C,char[][] board,String word,int index,boolean[][] visited){
        if(index==word.length()-1 && word.charAt(index)==board[s][e]){
            return true;
        }
        visited[s][e]=true;
        int diff[][]={{1,0},{0,-1},{-1,0},{0,1}};
        for(int i=0;i<4;i++){
            int adjr=diff[i][0]+s;
            int adjc=diff[i][1]+e;
            if(adjr>=0 && adjr<R && adjc>=0 && adjc<C && !visited[adjr][adjc] && board[adjr][adjc]==word.charAt(index+1)){
                if(search(adjr,adjc,R,C,board,word,index+1,visited)){
                    return true;
                }
            }
        }
        visited[s][e]=false;
        return false;

    }
}